### PR TITLE
[WIP] auto detect available services before registering all in DI

### PR DESCRIPTION
**Pull request by docteurklein**

### CODE REVIEW

File: DependencyInjection/KnpRadExtension.php [hunk 1]
```diff
@@ -6,20 +6,59 @@
...
+        }
+        if (!isset($bundles['TwigBundle'])) {
+            $config['mailer']['message_factory'] = false;
+            $config['flashes']['enabled'] = false;
+        }
+        if (!isset($bundles['SwiftmailerBundle'])) {
+            $config['mailer']['message_factory'] = false;
+        }
+
+        $frameworkConfigs = $container->getExtensionConfig($this->getAlias());
```
Comment (docteurklein) [inline DependencyInjection/KnpRadExtension.php:41]: I'm not sure this is really needed to put it in prepend method.  We could move all this to the `load` method maybe.

**Comment by lsmith77** [comment]
i guess this approach is the quickest to implement, but binds features to certain Bundles rather than certain services. then again, users can still override this automatic configuration in case they for some reason define the relevant service some other way.

**Comment by docteurklein** [comment]
indeed. Moreover, I don't like to get the config of frameworkBundle to enable/disable some things.

So indeed, the best would be to detect service existence. What would be the best approach ?

**Comment by lsmith77** [comment]
to be able to do that you will instead need to write a compiler pass. in this compiler pass you will then have to check if the service exists and if not disable certain features or throw exceptions to explain the necessary settings.

**Comment by docteurklein** [comment]
I was scared of this response :)
Sad that the extension configures all these services, and just after a compiler pass removes them.
It's far more complicated to list all services to remove than avoid loading a set of isolated services (separated independant yml files).
Moreover, we'll have to maintain the list of disableable services in both yml and compiler pass.

I don't know any other solution, so, ok for compiler pass!

File: DependencyInjection/KnpRadExtension.php (hunk 1 above)
Comment (gquemener) [inline DependencyInjection/KnpRadExtension.php:41]: Huh, and you can't do `$container->hasDefinition('some_service')` here?

Comment (docteurklein) [inline DependencyInjection/KnpRadExtension.php:41]: nope. $container is an isolated container that is available **only** for this extension. A compiler pass will then merge all the different containers into one big. The only thing that is shared is ~~aliases~~ parameters.

Comment (docteurklein) [inline DependencyInjection/KnpRadExtension.php:41]: https://github.com/symfony/symfony/blob/master/src/Symfony/Component/DependencyInjection/Compiler/MergeExtensionConfigurationPass.php#L46

Comment (gquemener) [inline DependencyInjection/KnpRadExtension.php:41]: m'kay

**Comment by docteurklein** [comment]
**bad idea of the day**: we could maybe tag the services with some aliases that would permit to know which should be disabled in a special case:

```
my.service:
    class: stdClass
    tags:
        - { name: 'remove-me-when', alias: no-doctrine }
        - { name: 'remove-me-when', alias: no-security }
        ### OR
        - { name: doctrine-required }
        ### OR:
        - { name: remove-if-no-service, service: doctrine.orm.entity_manager }
```

**Comment by lsmith77** [comment]
not such a bad idea .. there isnt really any performance penalty for tags and it would make it possible to make the compiler pass more generic. ie. reducing the risk that changes in our services break the behavior in the compiler pass.

**Comment by docteurklein** [comment]
yes. I would opt for 3rd syntax. EDIT: mmh, should the service definition know exactly which service it's (potentially indirectly) related to ? Or should we prefer give more abstract names (f.e: no-doctrine) and delegate this to the compiler pass ?

**Comment by gquemener** [comment]
I see some redundance in this tag idea:

``` yaml
my.service
    class: ...
    arguments:
        - @some.service
        - @another.service
    tags:
        - { name: 'remove-me-when', service: some.service }
        - { name: 'remove-me-when', service: another.service }
```

Couldn't it be possible to use directly the arguments and check existence of them?

**Comment by docteurklein** [comment]
@gquemener I think no, because it's potentially not a direct dependency to the missing service, but one related to it.
Still, we could maybe cascade the deletion, but is this a good idea? 
Just for you to know, there is already the information of missing service, with `on-invalid: ignore`.

**Pull request closed without merging**